bookkeeper-server/src/main/java/org/apache/bookkeeper/client/EnsemblePlacementPolicy.java

```diff
@@ -294,33 +294,39 @@ public BookieSocketAddress replaceBookie(int ensembleSize,
      *
      * @param ensemble
      *          Ensemble to read entries.
-     * @param writeSet
-     *          Write quorum to read entries.
      * @param bookieFailureHistory
      *          Observed failures on the bookies
-     * @return read sequence of bookies
+     * @param writeSet
+     *          Write quorum to read entries. This will be modified, rather than
+     *          allocating a new WriteSet.
+     * @return The read sequence. This will be the same object as the passed in
+     *         writeSet.
      * @since 4.5
      */
-    public List<Integer> reorderReadSequence(ArrayList<BookieSocketAddress> ensemble,
-                                             List<Integer> writeSet,
-                                             Map<BookieSocketAddress, Long> bookieFailureHistory);
+    public DistributionSchedule.WriteSet reorderReadSequence(
+            ArrayList<BookieSocketAddress> ensemble,
+            Map<BookieSocketAddress, Long> bookieFailureHistory,
+            DistributionSchedule.WriteSet writeSet);
 
 
     /**
      * Reorder the read last add confirmed sequence of a given write quorum <i>writeSet</i>.
      *
      * @param ensemble
      *          Ensemble to read entries.
-     * @param writeSet
-     *          Write quorum to read entries.
      * @param bookieFailureHistory
      *          Observed failures on the bookies
-     * @return read sequence of bookies
+     * @param writeSet
+     *          Write quorum to read entries. This will be modified, rather than
+     *          allocating a new WriteSet.
+     * @return The read sequence. This will be the same object as the passed in
+     *         writeSet.
      * @since 4.5
      */
-    public List<Integer> reorderReadLACSequence(ArrayList<BookieSocketAddress> ensemble,
-                                                List<Integer> writeSet,
-                                                Map<BookieSocketAddress, Long> bookieFailureHistory);
+    public DistributionSchedule.WriteSet reorderReadLACSequence(
```

Comment:
we could use 'void' as return type:
we are changing the signature and we do not want client code to allocate WriteSet but we would like the given WriteSet to be modified.

Anyway I am fine with this version.
But I would like to know @ivankelly  and @sijie options

Reply:
@eolivelli I think void would make it a little harder to use.
For example, in a lot of places we do
```
WriteSet reordered = policy.reorderReadSequnce(ensemble, failed, schedule.getWriteSet(eid))
```
If reorder returned void, then we would need to first assign the write set returned from getWriteSet to a variable and then pass it in.

Reply:
@ivankelly  I see, thank you for your clarification.